kube-runtime/src/finalizer.rs

```diff
@@ -207,3 +208,50 @@ pub enum Event<K> {
     /// - The grinch's heart grows a size or two
     Cleanup(K),
 }
+
+#[derive(Debug, Snafu)]
+pub enum DeleteError {
+    #[snafu(display("failed to delete object: {}", source))]
+    DeleteFailed { source: kube::Error },
+    #[snafu(display("failed to probe for whether the object is deleted yet: {}", source))]
+    DeleteProbeFailed { source: watcher::Error },
+}
+
+/// Try to delete an object, and wait for it to disappear from the Kubernetes API.
+/// If you do not wish to wait for the object to be deleted then you should use
+/// [`Api::delete`] instead.
+///
+/// This means that this function will not complete until all finalizers are done.
+///
+/// # Errors
+///
+/// This function fails if the deletion or status probe failed. In particular, it requires
+/// permission to use the following Kubernetes verbs:
+///
+/// - delete
+/// - list
+/// - watch
+///
+/// Note that the deletion will continue in the background if the probe fails.
+pub async fn finalize_and_delete<K>(api: &Api<K>, name: &str, dp: &DeleteParams) -> Result<(), DeleteError>
+where
+    K: Clone + DeserializeOwned + Debug + Send + Resource + 'static,
+{
+    api.delete(name, dp).await.context(DeleteFailed)?;
+    let watcher = watcher(api.clone(), ListParams {
+        field_selector: Some(format!("metadata.name={}", name)),
+        ..Default::default()
+    });
+    pin_mut!(watcher);
+    while let Some(event) = watcher.try_next().await.context(DeleteProbeFailed)? {
```

Comment:
What'd happen here if the object doesn't exist (anymore) between the `delete` call and the beginning of the watch?

Reply:
Then the `Restarted([])`  case fires, and we stop the watch immediately.